[ALLUXIO-2021] Keep Alluxio and UFS (local&hdfs) files user/group/permission in sync

ifcharming opened the pull request:
Addresses https://alluxio.atlassian.net/browse/ALLUXIO-2021

This PR:
- Adds new Set and Get user/group/permission in UFS interface.
- Implements user/group/permission propagation for local and HDFS. Not supported for Object storages.
- Create and complete UFS files now supports specifying the `PermissionStatus`
- When UFS delegation (decoupling) is turned on, permission is added to the thrift API. Therefore, worker would be able to know the end-user identify and permission to set for the file being created.
- Add `permissionStatus` in `persistFile` to propagate Alluxio ACL to UFS.
- Make `loadMetadata` (for both dir and files) load UFS ACL to Alluxio namespace.
- chown/chgrp/chmod ops are propagated to UFS if the Alluxio file is persisted

Tested locally with 
- different UFSs: Local, HDFS, secure HDFS
- `alluxio.user.ufs.delegation.enabled` on/off 

Future work:
- Add more unit test coverage
- Add integration test for ACL sync between Alluxio and UFS.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10389/<h2>Build result: FAILURE</span></h2>[...truncated 943 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.2.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.2.0-SNAPSHOT/alluxio-core-common-1.2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.2.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.2.0-SNAPSHOT/alluxio-core-common-1.2.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.2.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.2.0-SNAPSHOT/alluxio-core-common-1.2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/s3/pom.xml to org.alluxio/alluxio-underfs-s3/1.2.0-SNAPSHOT/alluxio-underfs-s3-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/assembly/pom.xml to org.alluxio/alluxio-assemblies/1.2.0-SNAPSHOT/alluxio-assemblies-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.2.0-SNAPSHOT/alluxio-shell-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/pom.xml to org.alluxio/alluxio-underfs-oss/1.2.0-SNAPSHOT/alluxio-underfs-oss-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.2.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-oss/1.2.0-SNAPSHOT/alluxio-underfs-oss-1.2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/oss/target/alluxio-underfs-oss-1.2.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-oss/1.2.0-SNAPSHOT/alluxio-underfs-oss-1.2.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/pom.xml to org.alluxio/alluxio-keyvalue-client/1.2.0-SNAPSHOT/alluxio-keyvalue-client-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.2.0-SNAPSHOT/alluxio-underfs-hdfs-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.2.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.2.0-SNAPSHOT/alluxio-underfs-hdfs-1.2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.2.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.2.0-SNAPSHOT/alluxio-underfs-hdfs-1.2.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.2.0-SNAPSHOT/alluxio-underfs-swift-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.2.0-SNAPSHOT/alluxio-examples-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.2.0-SNAPSHOT/alluxio-core-server-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.2.0-SNAPSHOT/alluxio-keyvalue-common-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.2.0-SNAPSHOT/alluxio-underfs-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.2.0-SNAPSHOT/alluxio-underfs-glusterfs-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.2.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.2.0-SNAPSHOT/alluxio-underfs-glusterfs-1.2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.2.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.2.0-SNAPSHOT/alluxio-underfs-glusterfs-1.2.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.2.0-SNAPSHOT/alluxio-integration-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.2.0-SNAPSHOT/alluxio-keyvalue-1.2.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10390/
Test PASSed.

jsimsa commented on core/client/src/main/java/alluxio/client/file/FileOutStream.java:
```diff
@@ -116,7 +119,7 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options) throws IOExcepti
         UnderFileSystem ufs = UnderFileSystem.get(tmpPath, ClientContext.getConf());
         // TODO(jiri): Implement collection of temporary files left behind by dead clients.
-        mUnderStorageOutputStream = ufs.create(tmpPath, (int) mBlockSize);
+        mUnderStorageOutputStream = ufs.create(tmpPath, options.getPermissionStatus());
```
why is block size removed from the `create` call?

ifcharming replied:
Corrected with using `UnderFileSystemCreateOptions`.
Previously because the `BlockSize` is not actually consumed in UFS `create()`. See the `HdfsUnderFileSystem` implementation for details.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/CompleteUfsFileOptions.java:
```diff
@@ -32,32 +35,37 @@
+      // TODO(chaomin): set permission based on the alluxio file. Not needed for now since the
+      // file is always created with default permission.
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
why do we convert the IOException to a RuntimeException?

ifcharming replied:
Discussed offline. Throw IOException for `defaults()`.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/CreateUfsFileOptions.java:
```diff
@@ -11,28 +11,103 @@
+      // TODO(chaomin): set permission based on the alluxio file. Not needed for now since the
+      // file is always created with default permission.
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
ditto, why is the exception converted from IOException to RuntimeException?

ifcharming replied:
Discussed offline. Throw IOException for `defaults()`.

ifcharming replied:
Discussed offline. Throw IOException for `defaults()`.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/OutStreamOptions.java:
```diff
@@ -57,6 +61,13 @@ private OutStreamOptions() {
+      mPermissionStatus = PermissionStatus.defaults().setUserFromLoginModule(conf)
+          .applyFileUMask(conf);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
ditto

jsimsa replied:
if the intention is not to catch the exception thrown by `setUserFromLoginModule`, then perhaps the method (as opposed to its callsites) should be doing the conversion

ifcharming replied:
This case is a bit different, I'm logging a warning message instead of changing the caller site too much.

jsimsa commented on core/client/src/test/java/alluxio/client/file/options/OutStreamOptionsTest.java:
```diff
@@ -21,8 +21,10 @@
 import alluxio.client.file.policy.LocalFirstPolicy;
```
is there no unit test for CreateUfsFileOptions / CompleteUfsFileOptions? if not, could you please add one and make sure your changes are included in the test? thanks

ifcharming replied:
New unit tests added.

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -115,7 +125,21 @@ public int getValue() {
-      Key key = new Key(new AlluxioURI(path));
+      PermissionStatus ps = PermissionStatus.defaults();
+      try {
+        // TODO(chaomin): consider adding a JVM-level constant to distinguish between Alluxio server
```
@calvinjia @apc999 @gpang is there an idiomatic way to determine if the current process is Alluxio master vs. Alluxio worker vs. Alluxio client?

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -354,6 +400,16 @@ public void configureProperties() throws IOException {
+   * @return A {@code OutputStream} object
+   * @throws IOException if a non-Alluxio error occurs
+   */
+  public abstract OutputStream create(String path, PermissionStatus ps) throws IOException;
```
I think at this point it would make sense to switch to provide two create methods:

`public abstract OutputStream create(String path)`

and

`public abstract OutputStream(String path, CreateOptions options)`

where options contains, permission status, block size (in bytes), and replication

ifcharming replied:
Thanks for the suggestion. Yes, this is for sure better.
I discussed with Calvin earlier and thought there will be a cleanup for UFS API anyways, so I didn't change in this PR. Since you raise it up, I just updated the PR with the change.

jsimsa commented on core/common/src/main/java/alluxio/util/io/FileUtils.java:
```diff
@@ -98,7 +138,7 @@ public static void changeLocalFileUser(String path, String user) throws IOExcept
         Files.getFileAttributeView(Paths.get(path), PosixFileAttributeView.class,
             LinkOption.NOFOLLOW_LINKS);
-    UserPrincipal userPrincipal = lookupService.lookupPrincipalByGroupName(user);
+    UserPrincipal userPrincipal = lookupService.lookupPrincipalByName(user);
```
was this a bug?

ifcharming replied:
Yes, this was a bug.

jsimsa replied:
thanks for the fix!

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -627,8 +699,36 @@ public void setProperties(Map<String, String> properties) {
+   * @return the owner of the file
+   * @throws IOException if a non-Alluxio error occurs
+   */
+  public abstract String getOwner(String path) throws IOException;
```
would it make sense to have provide getPermissionStatus and setPermissionStatus instead of the individual owner (we should be using either owner or user everywhere), group, and permission methods?

on a related note @apc999 :

I think that what we refer to in our code base as permission (e.g. 0700) is typically referred to mode (see for example man page for chmod). I think it would make more sense to use permission to describe the combination user, group, and mode together. So instead of PermissionStatus we would just have Permission, which contains user, group, and mode.

ifcharming replied:
Thanks for pointing this out. Once your [rename PR] (https://github.com/Alluxio/alluxio/pull/3562/files) is merged, I'll updated this one.

ifcharming replied:
Rename PR merged.
I think it makes sense to keep the different interface for UFS owner/group/mode setter and getters. Because Permission is only a concept in Alluxio space, and in many cases we only want to update part of Permission to UFS files.

jsimsa replied:
sgtm

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2460,8 +2507,14 @@ private void setAttributeFromEntry(SetAttributeEntry entry) throws FileDoesNotEx
+      try {
+        setAttributeInternal(inodePath, entry.getOpTimeMs(), options);
+      } catch (FileDoesNotExistException | InvalidPathException | AccessControlException e) {
+        throw new IOException("Failed to setAttributeFromEntry", e);
```
why are we converting the exceptions here?

ifcharming replied:
Done.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2460,8 +2507,14 @@ private void setAttributeFromEntry(SetAttributeEntry entry) throws FileDoesNotEx
+      }
       // Intentionally not journaling the persisted inodes from setAttributeInternal
+    } catch (FileDoesNotExistException e) {
+      throw new IOException("Failed to setAttributeFromEntry", e);
```
ditto

ifcharming replied:
Done, updated, thanks.

jsimsa commented on core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java:
```diff
@@ -258,23 +261,26 @@ private InputStream openAtPosition(long position) throws IOException {
      */
-    private OutputStreamAgent(long sessionId, long agentId, AlluxioURI ufsUri, Configuration conf)
+    private OutputStreamAgent(long sessionId, long agentId, AlluxioURI ufsUri, Configuration conf,
+                              PermissionStatus ps)
```
weird indentation

ifcharming replied:
done.

jsimsa commented on tests/src/test/java/alluxio/hadoop/FileSystemAclIntegrationTest.java:
```diff
@@ -33,6 +34,7 @@
  * Integration tests for {@link FileSystem#setOwner(Path, String, String)} and
  * {@link FileSystem#setPermission(Path, org.apache.hadoop.fs.permission.FsPermission)}.
  */
+@Ignore("TODO(chaomin): try to fix this because setOwner/setGroup now is propogated to UFS.")
```
remove "try to" :-)

ifcharming replied:
done.

jsimsa commented on tests/src/test/java/alluxio/master/JournalIntegrationTest.java:
```diff
@@ -537,6 +538,7 @@ private void rawTableTestUtil(FileInfo fileInfo) throws Exception {
   }
 
   @Test
+  @Ignore("TODO(chaomin): try to fix this because setOwner/setGroup now is propogated to UFS.")
```
ditto

ifcharming replied:
done.

jsimsa commented on underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:
```diff
@@ -117,15 +118,23 @@ public OutputStream create(String path) throws IOException {
   @Override
   public OutputStream create(String path, int blockSizeByte) throws IOException {
-    LOG.warn("blocksize is not supported with OSSUnderFileSystem. Block size will be ignored.");
+    LOG.debug("blocksize is not supported with OSSUnderFileSystem. Block size will be ignored.");
```
why are you changing the LOG level?

ifcharming replied:
yes, this was intended. See ALLUXIO-2017, ALLUXIO-2018.
Now those are all removed anyways.

jsimsa commented on underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:
```diff
@@ -317,7 +334,25 @@ public void setOwner(String path, String user, String group) {}
+  // Not supported
+  @Override
+  public short getPermission(String path) throws IOException {
+    return (short) 0777;
```
should the be using `Constants.DEFAULT_FS_FULL_PERMISSION`

ifcharming replied:
done.

jsimsa commented on underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java:
```diff
@@ -343,7 +360,25 @@ public void setOwner(String path, String user, String group) {}
+  // Not supported
+  @Override
+  public short getPermission(String path) throws IOException {
+    return (short) 0777;
```
should the be using `Constants.DEFAULT_FS_FULL_PERMISSION`

ifcharming replied:
done.

jsimsa commented on core/common/src/main/java/alluxio/util/io/FileUtils.java:
```diff
@@ -86,6 +87,45 @@ public static void changeLocalFileToFullPermission(String filePath) throws IOExc
+   * @return the file permission in posix format, e.g. "rwxr--r--"
+   * @throws IOException when fails to get the permission
+   */
+  public static String getLocalFilePermission(String filePath) throws IOException {
```
the only consumer than converts this to mode so I think this method should instead do the conversion

jsimsa replied:
simply iterating over the contents of `Files.readAttributes(Paths.get(filePath), PosixFileAttributes.class);` and computing the correct `short`

ifcharming replied:
done, thanks

jsimsa commented on underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -384,7 +401,25 @@ public void setOwner(String path, String user, String group) {}
+  // Not supported
+  @Override
+  public short getPermission(String path) throws IOException {
+    return (short) 0777;
```
should this be using `Constants.DEFAULT_FS_FULL_PERMISSION`?

ifcharming replied:
done.

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java:
```diff
@@ -343,7 +360,25 @@ public void setOwner(String path, String user, String group) {}
+  // Not supported
+  @Override
+  public short getPermission(String path) throws IOException {
+    return (short) 0777;
```
should this be using `Constants.DEFAULT_FS_FULL_PERMISSION`?

ifcharming replied:
done.

jsimsa commented:
@ifcharming thanks for the PR, took a pass

the following comments apply generally:
- use named constants (e.g. `Constants.DEFAULT_FS_FULL_PERMISSION`) instead of unnamed ones (e.g. 0777)
- do not convert exceptions at will
- use the options idiom for methods that can take a different set of options (`create` and `mkdirs` in UnderFileSystem interface)

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10399/
Test PASSed.

ifcharming commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -560,6 +616,22 @@ public abstract OutputStream create(String path, short replication, int blockSiz
+   *         otherwise
+   * @throws IOException if a non-Alluxio error occurs
+   */
+  public abstract boolean mkdirs(String path, boolean createParent, PermissionStatus ps)
```
Done, same as above.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10400/
Test PASSed.

ifcharming commented:
thanks a lot @jsimsa for taking a look. I've addressed most of your comments. Can you PTAL?
- Updated `create` and `mkdirs` with options in UnderFileSystem interface.
- Better exception handling
- Unit tests added for `options` and local FileUtils.
- Nits fixes

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10401/
Test PASSed.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/CompleteUfsFileOptions.java:
```diff
@@ -32,32 +34,35 @@
   private final String mUser;
```
should we be using owner?

ifcharming replied:
I was planing to do the renaming from all `user` to `owner` in a separate PR.

jsimsa replied:
FYI, https://github.com/Alluxio/alluxio/pull/3566

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/CreateUfsFileOptions.java:
```diff
@@ -11,28 +11,100 @@
 @NotThreadSafe
 public final class CreateUfsFileOptions {
+  /** The ufs user this file should be owned by. */
+  private final String mUser;
```
`mOwner`?

jsimsa commented on core/client/src/test/java/alluxio/client/file/options/CompleteUfsFileOptionsTest.java:
```diff
@@ -0,0 +1,50 @@
+    Assert.assertEquals(expectedPs.getPermission().toShort(), options.getPermission());
+    ClientTestUtils.resetClientContext();
+  }
+}
```
please also add a "fieldsTest"  (test setters and getters) and "equalsTest" (using `CommonTestUtils.testEquals`)

ifcharming replied:
done.

jsimsa commented on core/client/src/test/java/alluxio/client/file/options/CreateUfsFileOptionsTest.java:
```diff
@@ -0,0 +1,50 @@
+    Assert.assertEquals(expectedPs.getPermission().toShort(), options.getPermission());
+    ClientTestUtils.resetClientContext();
+  }
+}
```
please also add a "fieldsTest"  (test setters and getters) and "equalsTest" (using `CommonTestUtils.testEquals`)

ifcharming replied:
done.

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -140,16 +166,19 @@ UnderFileSystem get(String path, Object ufsConf, Configuration configuration) {
     private final String mScheme;
     private final String mAuthority;
-    // TODO(peis): Add UGI information.
+    private final String mUser;
```
owner?

ifcharming replied:
done, thanks for your refactoring PR.

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -15,10 +15,15 @@
 import alluxio.Constants;
 import alluxio.collections.Pair;
+import alluxio.security.authorization.PermissionStatus;
+import alluxio.underfs.options.UnderFileSystemCreateOptions;
```
I think we can call the options just `CreateOptions`

ifcharming replied:
updated. I was trying to distinguish UFS op option names from Alluxio op options.

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -15,10 +15,15 @@
 import alluxio.collections.Pair;
+import alluxio.security.authorization.PermissionStatus;
+import alluxio.underfs.options.UnderFileSystemCreateOptions;
+import alluxio.underfs.options.UnderFileSystemMkdirsOptions;
```
ditto, just `MkdirsOptions`

ifcharming replied:
renamed.

jsimsa commented on core/common/src/main/java/alluxio/underfs/options/UnderFileSystemCreateOptions.java:
```diff
@@ -0,0 +1,93 @@
+  // Permission status to set for the file being created.
+  private PermissionStatus mPermissionStatus;
+  // Block size in bytes for HDFS files only, not applicable for other Under FileSystems.
+  private long mBlockSizeByte;
```
if this field is actually not used, there is no point adding it (similarly to how you didn't add the `replication` field)

ifcharming replied:
done, thx

jsimsa commented on core/common/src/main/java/alluxio/util/io/FileUtils.java:
```diff
@@ -86,6 +89,51 @@ public static void changeLocalFileToFullPermission(String filePath) throws IOExc
+   * @return the file permission in short, e.g. 0777
+   * @throws IOException when fails to get the permission
+   */
+  public static short getLocalFilePermission(String filePath) throws IOException {
```
once my PR goes in, this should be call `getLocalFileMode`

ifcharming replied:
done.

jsimsa commented on core/common/src/test/java/alluxio/underfs/options/UnderFileSystemCreateOptionsTest.java:
```diff
@@ -0,0 +1,88 @@
+    Assert.assertEquals(blockSize, options.getBlockSizeByte());
+    Assert.assertEquals(ps, options.getPermissionStatus());
+  }
+}
```
please add "equalsTest"

ifcharming replied:
done.

jsimsa commented on core/common/src/test/java/alluxio/underfs/options/UnderFileSystemMkdirsOptionsTest.java:
```diff
@@ -0,0 +1,86 @@
+    Assert.assertEquals(createParent, options.getCreateParent());
+    Assert.assertEquals(ps, options.getPermissionStatus());
+  }
+}
```
please add "equalsTest"

ifcharming replied:
done.

jsimsa commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -232,4 +232,31 @@ public void createDirTest() throws IOException {
+   * Tests the {@link FileUtils#getLocalFilePermission(String)}} method.
+   */
+  @Test
+  public void getLocalFilePermissionTest() throws IOException {
```
once my PR goes in this should be called `getLocalFileModeTest`

ifcharming replied:
done.

jsimsa commented on core/common/src/thrift/file_system_worker.thrift:
```diff
@@ -12,9 +12,13 @@ struct CloseUfsFileTOptions {
 struct CreateUfsFileTOptions {
+  1: optional string user
+  2: optional string group
+  3: optional i16 permission
```
ditto (s/permission/mode after my PR goes in)

ifcharming replied:
done.

jsimsa commented on core/common/src/thrift/file_system_worker.thrift:
```diff
@@ -12,9 +12,13 @@ struct CloseUfsFileTOptions {
 struct CompleteUfsFileTOptions {
   1: optional string user
```
we should probably replace user with owner everywhere, but it might make sense to do that in a separate PR (for instance, we can't simply rename this field because it would break backwards compatibility ... so we will need to add a new one a support both until support for `user` is removed)

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2438,7 +2488,8 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
    * @param entry the entry to use
    * @throws FileDoesNotExistException if the file does not exist
```
javadoc exceptions

ifcharming replied:
done.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2389,7 +2409,7 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
    */
```
javadoc exceptions

ifcharming replied:
done.

jsimsa commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1845,7 +1846,7 @@ private long loadMetadataAndJournal(LockedInodePath inodePath, LoadMetadataOptio
   private long loadFileMetadataAndJournal(LockedInodePath inodePath,
```
javadoc exceptions

ifcharming replied:
done.

jsimsa commented on core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java:
```diff
@@ -258,23 +262,27 @@ private InputStream openAtPosition(long position) throws IOException {
      * @throws IOException if an error occurs when interacting with the UFS
      */
-    private OutputStreamAgent(long sessionId, long agentId, AlluxioURI ufsUri, Configuration conf)
+    private OutputStreamAgent(
```
the indentation here seems odd

ifcharming replied:
done.

jsimsa commented on core/common/src/main/java/alluxio/underfs/options/UnderFileSystemCreateOptions.java:
```diff
@@ -0,0 +1,93 @@
+    // file creation.
+    mPermissionStatus = PermissionStatus.defaults().applyFileUMask(conf);
+    // By default HDFS block size is 64MB.
+    mBlockSizeByte = 64 * Constants.MB;
```
instead of duplicating logic between `CreateOptions()` and `CreateOptions(conf)`, `CreateOptions()` should simply invoked `CreateOptions(context.getConf())` with the appropriate context

jsimsa commented on core/common/src/main/java/alluxio/underfs/options/UnderFileSystemMkdirsOptions.java:
```diff
@@ -0,0 +1,92 @@
+    // directories creation.
+    mPermissionStatus = PermissionStatus.defaults().applyDirectoryUMask(conf);
+    // By default create parent is true.
+    mCreateParent = true;
```
ditto

jsimsa commented on core/common/src/main/java/alluxio/underfs/options/UnderFileSystemCreateOptions.java:
```diff
@@ -0,0 +1,93 @@
+  /**
+   * @return the default {@link UnderFileSystemCreateOptions}
+   */
+  public static UnderFileSystemCreateOptions defaults() {
```
it seems odd to provide `defaults()` and a constructor ... it is mixing two idioms

we should either have two constructors or two defaults (`defaults()` and `defaults(conf)`, which is what I prefer)

jsimsa replied:
ideally we would only have one `defaults()` method which fetched the configuration from the context ... this is how it works for all other options

ifcharming replied:
The reason to have the configuration as the param is that UFS module can't tell which component is actually calling UFS ops.

jsimsa commented on core/common/src/main/java/alluxio/underfs/options/UnderFileSystemMkdirsOptions.java:
```diff
@@ -0,0 +1,92 @@
+  /**
+   * @return the default {@link UnderFileSystemMkdirsOptions}
+   */
+  public static UnderFileSystemMkdirsOptions defaults() {
```
it seems odd to provide defaults() and a constructor ... it is mixing two idioms

we should either have two constructors or two defaults (defaults() and defaults(conf), which is what I prefer)

jsimsa replied:
ideally we would only have one `defaults()` method which fetched the configuration from the context ... this is how it works for all other options

jsimsa commented on underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java:
```diff
@@ -115,29 +117,18 @@ public void connectFromWorker(Configuration conf, String hostname) {
+
+  @Override
+  public OutputStream create(String path, UnderFileSystemCreateOptions options) throws IOException {
+    // Block size and permission status in options are ignored when creating GCS object.
```
This comment will get stale when new fields are added to `options`. I would simply remove it.

ifcharming replied:
done.

jsimsa commented on underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java:
```diff
@@ -128,12 +131,35 @@ public void close() throws IOException {
+        LOG.debug("Creating HDFS file at {} with perm {}", path, ps.toString());
+        return FileSystem.create(mFileSystem, new Path(path),
+            new FsPermission(ps.getPermission().toShort()));
+        // TODO(hy): Fix this.
```
no need to preserve the commented out code ... instead add a method-level TODO to support block size and replication

ifcharming replied:
done.

jsimsa commented on underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:
```diff
@@ -110,6 +112,12 @@ public void close() throws IOException {
+
+  @Override
+  public OutputStream create(String path, UnderFileSystemCreateOptions options) throws IOException {
+    // Block size and permission status in options are ignored when creating OSS object.
```
This comment will get stale when new fields are added to `options`. I would simply remove it.

ifcharming replied:
done.

jsimsa commented on underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -155,29 +157,18 @@ public void connectFromWorker(Configuration conf, String hostname) {
+
+  @Override
+  public OutputStream create(String path, UnderFileSystemCreateOptions options) throws IOException {
+    // Block size and permission status in options are ignored when creating S3 object.
```
This comment will get stale when new fields are added to `options`. I would simply remove it.

ifcharming replied:
done.

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java:
```diff
@@ -133,6 +135,12 @@ public void connectFromWorker(Configuration conf, String hostname) {
+
+  @Override
+  public OutputStream create(String path, UnderFileSystemCreateOptions options) throws IOException {
+    // Block size and permission status in options are ignored when creating Swift object.
```
This comment will get stale when new fields are added to `options`. I would simply remove it.

ifcharming replied:
done.

jsimsa commented:
@ifcharming thanks for addressing my comments, did another pass

jsimsa commented:
@ifcharming FYI my authorizaton rename PR has been merged

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10410/
Test FAILed.

ifcharming commented on core/common/src/thrift/file_system_worker.thrift:
```diff
@@ -12,9 +12,13 @@ struct CloseUfsFileTOptions {
 struct CompleteUfsFileTOptions {
   1: optional string user
   2: optional string group
+  3: optional i16 permission
```
done.

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -627,8 +699,36 @@ public void setProperties(Map<String, String> properties) {
    * @throws IOException if a non-Alluxio error occurs
    */
-  public abstract void setPermission(String path, String posixPerm) throws IOException;
+  public abstract short getPermission(String path) throws IOException;
```
this should be `getMode`

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -627,8 +699,36 @@ public void setProperties(Map<String, String> properties) {
+   * @param permission the permission to set in short format, e.g. 0777
+   * @throws IOException if a non-Alluxio error occurs
+   */
+  public abstract void setPermission(String path, short permission) throws IOException;
```
`setMode`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10415/
Test PASSed.

jsimsa commented:
module re-enabling the ignored tests, this looks good to me

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10418/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10416/
Test PASSed.

ifcharming commented:
@apc999 @calvinjia  can you please take a look at this PR?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10420/
Test PASSed.

calvinjia commented on core/client/src/main/java/alluxio/client/file/FileOutStream.java:
```diff
@@ -116,7 +120,8 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options) throws IOExcepti
         // TODO(jiri): Implement collection of temporary files left behind by dead clients.
-        mUnderStorageOutputStream = ufs.create(tmpPath, (int) mBlockSize);
+        CreateOptions ufsCreateOptions = new CreateOptions().setPermission(options.getPermission());
+        mUnderStorageOutputStream = ufs.create(tmpPath, ufsCreateOptions);
```
Is it confusing that the varaible `ufsCreateOptions` of type `CreateOptions` and not `CreateUfsFileOptions`? Maybe just name the variable `createOptions`?

ifcharming replied:
done.

calvinjia commented on core/client/src/main/java/alluxio/client/file/FileSystemUtils.java:
```diff
@@ -161,7 +163,12 @@ public static long persistFile(FileSystem fs, AlluxioURI uri, URIStatus status,
+      // TODO(chaomin): should also propagate ancestor dirs permission to UFS.
+      URIStatus uriStatus = fs.getStatus(uri);
+      Permission perm = new Permission(uriStatus.getUserName(), uriStatus.getGroupName(),
+          (short) uriStatus.getPermission());
```
Perhaps we should change this to `getMode` and deprecate `getPermission`? Since it seems like we use the term permission for something else.

ifcharming replied:
Yes we should. FYI another PR https://github.com/Alluxio/alluxio/pull/3566 did the rename from permission->mode. I'll update once that is merged.

ifcharming replied:
Updated.

calvinjia commented on core/client/src/main/java/alluxio/client/file/FileOutStream.java:
```diff
@@ -116,7 +120,8 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options) throws IOExcepti
         UnderFileSystem ufs = UnderFileSystem.get(tmpPath, ClientContext.getConf());
         // TODO(jiri): Implement collection of temporary files left behind by dead clients.
-        mUnderStorageOutputStream = ufs.create(tmpPath, (int) mBlockSize);
+        CreateOptions ufsCreateOptions = new CreateOptions().setPermission(options.getPermission());
```
Why do we store the entire permission object here instead of separately storing owner/group/mode?

ifcharming replied:
I think it makes sense to use a unified `Permission` object, like what we did in `inodeTree`, instead of separately storing owner/group/mode.

calvinjia replied:
But for the other option we store the pieces individually right?

ifcharming replied:
Yup, the other options have to be converted to Thrift, so its better to keep `*Option` and `*TOption` same. While the UFS options here are not converted to Thrift options, so I used `Permission` object.
I'm open to alternative though. Let me know if you have a strong preference.

calvinjia replied:
Makes sense.

calvinjia commented on core/common/src/main/java/alluxio/underfs/options/CreateOptions.java:
```diff
@@ -0,0 +1,90 @@
+   */
+  public CreateOptions() {
+    mPermission = Permission.defaults();
+  }
```
Any reason why we want to deviate from the `CreateOptions.defaults()` scheme and go to `new CreateOptions()`?

calvinjia replied:
Makes sense.

jsimsa replied:
add a TODO to switch this to a single `defaults()` once / if we move to singleton configuration

calvinjia commented on core/client/src/main/java/alluxio/client/file/options/CompleteUfsFileOptions.java:
```diff
@@ -82,15 +97,69 @@ public boolean hasGroup() {
+   * @return if the mode has been set
+   */
+  public boolean hasMode() {
+    return mMode != Constants.INVALID_MODE;
```
Will this ever be false?

ifcharming replied:
not really, mode should always be set by default (to file Umask).

calvinjia commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -40,11 +45,18 @@
+  /** The user to access this {@link UnderFileSystem}. */
+  protected final String mUser;
+  /** The group to access this {@link UnderFileSystem}. */
+  protected final String mGroup;
```
How do we use these two values?

ifcharming replied:
it's not used yet...

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -1858,6 +1859,14 @@ private long loadFileMetadataAndJournal(LockedInodePath inodePath,
+      short ufsPermission = ufs.getMode(ufsUri.toString());
+      createFileOptions = createFileOptions.setPermission(
+          new Permission(ufsOwner, ufsGroup, ufsPermission));
+    }
```
A high level question: does it make sense to always populate the security info in the inode tree, but only do checking if security is enabled? That way if the user loads data and then turns on security, the permissions will appear instead of being default/empty.

ifcharming replied:
Good point. `isSecurityEnabled` is not necessary because we can still set permission on `inode` when security is disabled.
updated, thanks!

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2424,21 +2446,54 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
     }
     if (options.getPermission() != Constants.INVALID_MODE) {
       inode.setPermission(options.getPermission());
+      if (inode.isPersisted()) {
```
And the `if (inode.isPersisted())` check go outside the other `if` clauses?

ifcharming replied:
done, makes sense.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2424,21 +2446,54 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
+        }
+      }
     }
     if (options.getPermission() != Constants.INVALID_MODE) {
```
Confusion between `permission` and `mode`

ifcharming replied:
same as above, once #3566 is merged, there will no longer be confusion between `permission` and `mode`.

ifcharming replied:
Updated.

calvinjia commented:
@ifcharming 
Two high level points:
I think we should resolve the `mode/permission` naming terms. We should also consider having security metadata always updated and only checked against if we enable security.

ifcharming commented:
@calvinjia  thanks for taking a look! I've addressed your comments. Can you PTAL?

`mode/permission` confusion will be solved once #3566 is merged.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10432/
Test PASSed.

calvinjia commented:
@ifcharming Thanks, looks good overall, I had one followup question about consistently storing permission or owner/group/mode in options.

calvinjia commented:
LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10451/
Test PASSed.

ifcharming commented:
Merged https://github.com/Alluxio/alluxio/pull/3566, this PR is now ready for submission.
@apc999  can you please take a look? Thanks!

jsimsa commented on core/client/src/main/java/alluxio/client/file/FileOutStream.java:
```diff
@@ -116,7 +120,8 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options) throws IOExcepti
         UnderFileSystem ufs = UnderFileSystem.get(tmpPath, ClientContext.getConf());
         // TODO(jiri): Implement collection of temporary files left behind by dead clients.
-        mUnderStorageOutputStream = ufs.create(tmpPath, (int) mBlockSize);
+        CreateOptions createOptions = new CreateOptions().setPermission(options.getPermission());
```
is there any reason why `CreateUfsFileOptions` uses owner / group / mode fields while `CreateOptions` users permission field? would it make sense to just have a permission field for `CreateUfsFileOptions` as well?

ifcharming replied:
Because `CreateUfsFileOptions` and `CompleteUfsFileOptions` have to be converted to Thrift, so its better to keep *Option and *TOption same. While the UFS options here are not converted to Thrift options, so I used a single `Permission` object, which is cleaner and consistent with that in `InodeTree`

jsimsa replied:
I think it makes more sense for non-wire types to use `Permission` (similar to how `OutStreamOptions` does it) and do the packing / unpacking only when sending / receving things over the wire ... I don't feel strongly about this though so I will leave up to you

apc999 replied:
I kinda agree with @jsimsa here as I understand the purpose of using `xxxOptions` instead of `xxxTOptions` directly is to hide the specific way of sending data over wire, and easing the development is probably the final goal

ifcharming replied:
Make sense to me. I've updated to use `Permission` in `CreateUfsFileOptions` and `CompleteUfsFileOptions`.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/CompleteUfsFileOptions.java:
```diff
@@ -11,53 +11,50 @@
 public final class CompleteUfsFileOptions {
-  /** The ufs user this file should be owned by. */
-  private final String mOwner;
+  /** The ufs owner this file should be owned by. */
```
would it make sense to just have Permission field instead?

ifcharming replied:
Same reason as above.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10483/
Test PASSed.

jsimsa commented:
LGTM

apc999 commented on core/client/src/main/java/alluxio/client/file/options/CompleteUfsFileOptions.java:
```diff
@@ -11,53 +11,50 @@
-    }
-    String group = user;
-    return new CompleteUfsFileOptions(user, group);
+  public static CompleteUfsFileOptions defaults() throws IOException {
```
since the constructor here is nontrivial, better to javadoc what is the default here.
basically, telling people this default option value will contain owner group and mode from login module.

ifcharming replied:
added, thanks.

apc999 commented on core/client/src/test/java/alluxio/client/file/options/CompleteUfsFileOptionsTest.java:
```diff
@@ -0,0 +1,95 @@
+   */
+  @Test
+  public void defaultsTest() throws IOException {
+    ClientContext.getConf().set(Constants.SECURITY_AUTHENTICATION_TYPE, "SIMPLE");
```
`AuthType.SIMPLE.getAuthName()`?

ifcharming replied:
done.

apc999 commented on core/client/src/test/java/alluxio/client/file/options/CreateUfsFileOptionsTest.java:
```diff
@@ -0,0 +1,95 @@
+   */
+  @Test
+  public void defaultsTest() throws IOException {
+    ClientContext.getConf().set(Constants.SECURITY_AUTHENTICATION_TYPE, "SIMPLE");
```
ditto

ifcharming replied:
done.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10487/
Test PASSed.

apc999 commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -115,7 +127,21 @@ public int getValue() {
+        // TODO(chaomin): consider adding a JVM-level constant to distinguish between Alluxio server
+        // and client.
+        String loggerType = configuration.get(Constants.LOGGER_TYPE);
+        if (loggerType.equalsIgnoreCase("MASTER_LOGGER")
```
could you add a TODO here to tell whether we are in client JVM or not using some built-in util functions, rather than from logger types?

apc999 replied:
see above TODO. please discard my comment

apc999 commented:
LGTM. thanks for the awesome patch to make Alluxio more secure!
